aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp

```diff
@@ -474,84 +420,112 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_efficient_attention_nestedtensor_
+       max_seqlen_batch_q,
+       max_seqlen_batch_kv,
+       output_shape] = sdpa_nested_preprocessing(query, key, value);
 
-  TORCH_CHECK(
-      key.is_same(key) && query.is_same(value),
-      "Key and Value must be the same tensor");
-
-  int64_t Nnz_q = std::get<2>(cumulative_and_max_q_and_nnz_q);
-
-  // For the packed case we need to set the output size for dim 2 to 1
-  auto atten_size = get_nested_size_tensor(query).clone();
-  atten_size.index({at::indexing::Slice(), 1}) = 1;
-
-  auto qkv_buffer_reshaped = get_buffer(query)
-                                 .view({Nnz_q, 3, num_heads, head_dim})
-                                 .transpose(0, 1)
-                                 .contiguous();
-
-  auto q = qkv_buffer_reshaped[0];
-  auto k = qkv_buffer_reshaped[1];
-  auto v = qkv_buffer_reshaped[2];
-
-  TORCH_CHECK(q.is_contiguous());
-  TORCH_CHECK(k.is_contiguous());
-  TORCH_CHECK(v.is_contiguous());
-
-  // If we are passing in query, key, value all the same tensors then we have
-  // packed them into one tensor and need to slice for flash attention
-  Tensor attention =
-      std::get<0>(at::_flash_attention_forward(
-          q,
-          k,
-          v,
-          cumulative_sequence_length_q,
+  Tensor attention, log_sumexp, debug_attn_mask;
+  int64_t philox_seed{0}, philox_offset{0};
+  std::tie(attention, log_sumexp, philox_seed, philox_offset, debug_attn_mask) =
+      at::_flash_attention_forward(
+          query_buffer_reshaped,
+          key_buffer_reshaped,
+          value_buffer_reshaped,
           cumulative_sequence_length_q,
+          cumulative_sequence_length_k,
           max_seqlen_batch_q,
-          max_seqlen_batch_q,
+          max_seqlen_batch_kv,
           dropout_p,
           is_causal,
-          false));
-  // Output of flash_attention is a regular tensor lets wrap it back up to
-  // form a nested tensor
+          return_debug_mask);
+  // Reshape output to convert nnz to batch_size and seq_len
+  attention = wrap_buffer(attention.view(-1), output_shape).transpose(1, 2);
+  return std::make_tuple(
+      attention,
+      log_sumexp,
+      cumulative_sequence_length_q,
+      cumulative_sequence_length_k,
+      max_seqlen_batch_q,
+      max_seqlen_batch_kv,
+      philox_seed,
+      philox_offset,
+      debug_attn_mask);
+}
 
-  return wrap_buffer(attention.view(-1), atten_size);
+std::tuple<Tensor, Tensor>
+_scaled_dot_product_efficient_attention_nestedtensor_cuda(
+    const Tensor& query,
+    const Tensor& key,
+    const Tensor& value,
+    bool compute_log_sumexp,
+    bool is_causal) {
+  // c++17 structured bindings
+  auto
+      [query_buffer_reshaped,
+       key_buffer_reshaped,
+       value_buffer_reshaped,
+       cumulative_sequence_length_q,
+       cumulative_sequence_length_k,
+       max_seqlen_batch_q,
+       max_seqlen_batch_kv,
+       output_shape] = sdpa_nested_preprocessing(query, key, value);
+  TORCH_CHECK(max_seqlen_batch_q == max_seqlen_batch_kv)
```

Comment:
n00b question on my end, why do we need this TORCH_CHECK?

I've been running with max_seqlen_batch_q = 1 and  max_seqlen_batch_kv = smth else and it seemed to work

btw, what does [this part](https://github.com/pytorch/pytorch/blob/325b43661e44c99fb3a95bb2a4a94ad79f315c9d/aten/src/ATen/native/transformers/cuda/attention.cu#L909-L911) where it says max_seq_len_k is set in the kernel do (is it related?)